src/librustc_mir/borrow_check/nll/type_check/mod.rs

```diff
@@ -619,16 +619,35 @@ pub struct OutlivesSet<'tcx> {
 }
 
 #[derive(Copy, Clone, Debug)]
-pub struct Locations {
-    /// The location in the MIR that generated these constraints.
-    /// This is intended for error reporting and diagnosis; the
-    /// constraints may *take effect* at a distinct spot.
-    pub from_location: Location,
-
-    /// The constraints must be met at this location. In terms of the
-    /// NLL RFC, when you have a constraint `R1: R2 @ P`, this field
-    /// is the `P` value.
-    pub at_location: Location,
+pub enum Locations {
+    All,
```

Comment:
so, I'm torn. I was struck by your occasional use of `.unwrap_or(Location::START)`, because I wasn't clear on how that was justified. But then I looked at the code prior to this PR, and do see places where you have (quite naturally) replaced previous uses of START with All...

Anyway, reading over those cases led me back to this enum declaration. I thought at first that `All` is pretty self-documenting here in the enum declaration itself (i.e., it represents all locations, and it is used in cases where you are describing some constraint that is true everywhere in the control-flow graph, like type equivalences)...  But then after reading the comments that go with `Locations::Pair`, I sort of lost that perspective. I guess the main point is that some constraints are flow-independent throughout the function, and others are control-flow dependent, and that is the distinction encoded in this enum? Does that sound right?

But then ... how should one decide when it is appropriate to map `from_location()` (or `at_location()`) of `None` to `Location::START`? Is it just that since we're talking about an flow-independent constraint, then any location will do, so we might as well pick `START`?

Reply:
TBH, the code that "falls back" to START is a bit confusing. I'm not 100% convinced that is a correct thing to do, but I was just trying to keep existing behavior. In any case, START is a better choice for the old NLL because all other points are reachable from there. I can try to elaborate the comment a bit.

The problem I was specifically encountering here in the new analysis was that we would do this:

```
_0 = p
```

where `_0` is the return pointer. We treat `_0` like any other place, and hence we considered `_0` to not be live until that point! This meant that when we related `_0` to the return type at the entry, all of those subtype relations were lost. This is why I re-add it at each point. TBH it's .. maybe a hack? I wasn't fully satisfied.